modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EElvis.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.node;
+
+import org.elasticsearch.painless.AnalyzerCaster;
+import org.elasticsearch.painless.Definition;
+import org.elasticsearch.painless.Definition.Type;
+import org.elasticsearch.painless.Globals;
+import org.elasticsearch.painless.Locals;
+import org.elasticsearch.painless.Location;
+import org.elasticsearch.painless.MethodWriter;
+import org.objectweb.asm.Label;
+
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * The Elvis operator ({@code ?:}), a null coalescing operator. Binary operator that evaluates the first expression and return it if it is
+ * non null. If the first expression is null then it evaluates the second expression and returns it.
+ */
+public class EElvis extends AExpression {
```

Comment:
I spent a while thinking about this and I think I like the way I have it better. The thing that worries me is that we'd need to introduce a synthetic variable to keep from from evaluating the lhs twice. We can do that but it seems a bit complex and heavy. By heavy I mean that it relies on the jit to eliminate it. I expect it mostly can but the bytecode that EElvis emits is more obviously efficient to me.

One thing I'd like to do as a followup to this is to investigate the extra box/unbox step you get in code like `int x = foo?.x :? 0` when `foo.x` has return type `int`. I think the no-node approach or the node approach can optimize this away, I think. I feel like the approach I've started with is going to be easier/more obvious about it. I'm not even sure it is needed though. I'd have to run some comparisons to check first. I don't think that is a good argument for the node approach because it might not be needed at all and it might not be simpler but I wanted to type it out because it has been rattling around in my head.